### DISCUSSION TITLE
PromotionMgr.applyDiscounts(discountPlan) does not work

### DISCUSSION THREAD

**Discussion by Ivan-Melnyk** [SalesforceCommerceCloud/sfcc-ci#529] [Team Posts]
Hi Community,

I'm trying to apply PromotionMgr.applyDiscounts(discountPlan) for Basket, where I removed some Product Level discounts.

for example (just a sample):
var discountPlan = PromotionMgr.getDiscounts(Basket);  
for each (var pli in Basket.productLineItems) {  
  for each (var pd in discountPlan.getProductDiscounts(pli)) {  
    discountPlan.removeDiscount(pd);  
  }  
}  
PromotionMgr.applyDiscounts(discountPlan);  
Basket.updateTotals();
  
If I have any percentage Order level discount, after applying discounts in such way I still have wrong Basket.priceAdjustments[n].price value, like it was calculated based on the original product level discounts, like I've never removed product level discounts.

I've tried different Compatibility Mode of the sandbox, from 15.5 to 17.7 and it does not help.

Does anyone of you know how to solve it?

Thanks.

**Comment by tparikh2** [comment]
Hi Ivan,

Can you please explain the business case why you need to remove some
product discount ?

Can you use global exclusion rule in BM to exclude products from any
promotion?

Regards,
Tushar Parikh

**Comment by Ivan-Melnyk** [comment]
Hi @tparikh2 , yes, sure.
I'm working at Global-e, so we integrate our cartridge which do a lot of customization to how the price is rendered, and how the promotions are applied to the product/order, etc. One of such logic in our cartridge is to apply promotions (their discounts) to products which has "fixed prices". Fixed prices is not something that can be determined in BM or in Promotion configuration, but calculated in our scripts. so For example there are two products in the Basket, one of them has "fixed price", another one has "dynamic" price, so the Promotion should be applied only for "fixed price" line item, and not to the another one, so I'm iterating all products and their discounts in calculate.js and removing those discounts which should not be applied, after that I'm triggering PromotionMgr.applyDiscounts(discountPlan) with the discount plan which was modified, but it does not calculate then the Order level discounts properly (like in the example at the top), looks like SFCC did not realized that some of the product level discounts where removed, and it still calculates Order level discount with applying all product level discounts, including the removed ones.

**Comment by usmansystemsltd** [comment]
Ivan all you need to do is make sure the apply discount method with your customization is applied is the last one applied in calculatecart.js or else it will always bring back the old promotion plan with old discounts you removed other way around would be to create a function to remove discounts and where ever applydiscounts is called just call this function before it.

**Comment by Ivan-Melnyk** [comment]
@usmansystemsltd I've tried to apply that in different places and the result is still not what I expect.
Not sure which method of applyDisocunts(...) are you talking about, but I'm talking about applyDiscounts(discountPlan) with discountPlan, not Basket. I know that if you'll trigger applyDiscounts(Basket) all promotions and it's discounts will be set back to the basket, disregards of any customization, but that's different method.

**Comment by syed-shahjahan** [comment]
Hi @Ivan-Melnyk ,

Please try getting the available promotions after removing the promotions

var basketDiscounts = PromotionMgr.getDiscounts(Basket);
for each (var pli in Basket.productLineItems) {
for each (var pd in basketDiscounts.getProductDiscounts(pli)) {
basketDiscounts.removeDiscount(pd);
}
}

// Get available discounts after removing
    var discountsAfter = PromotionMgr.getDiscounts(basket, promotions);
//Apply available discounts
    PromotionMgr.applyDiscounts(discountsAfter);
Basket.updateTotals();

**Comment by Ivan-Melnyk** [comment]
Hi @syed-shahjahan thanks for advice,
but PromotionMgr.getDiscounts(Basket, PromotionPlan) will always return the initial discountPlan, before any customizationm, and I can't remove any promotion from PromotionPlan, since the promotion might be applicable to one product line item in the basket, and not for another one.

**Comment by tparikh2** [comment]
Hi Ivan,

You also need to recalculate tax using calculateTax(basket) after removing
product discount. You can do it before doing basket.updateTotals().

Either way, Is it possible to execute your script that will mark a product
as a fixed price or dynamic price ? If so then you can straight forward use
out of the box global promotion exclusion. You can execute this job along
with pricebook import so it will be always in synch.

Thanks & Regards,
Tushar Parikh.

+91-909-901-3316

**Comment by Ivan-Melnyk** [comment]
Hi @tparikh2 thanks for reply,
I'm calling basket.updateTotals() in the very end of the calculate.js, so it's not an issue.
It is not possible to mark the products anyhow by running the jobs, since we have about 200 countries and about 100 currencies, and each variant product might have fixed price for any combination of country + currency, so I need to handle it only in script. I was referring to this document https://documentation.b2c.commercecloud.salesforce.com/DOC2/topic/com.demandware.dochelp/DWAPI/scriptapi/html/api/class_dw_campaign_PromotionMgr.html?cp=0_16_2_2_24, and in theory it should work, but seems there is a bug in SFCC Core.
![image](https://user-images.githubusercontent.com/41744752/55402153-a4a92480-554a-11e9-804d-609bfb5180ef.png)

**Comment by GeneralPatten** [comment]
Ivan – can you provide an example of the promotion setup for your test case, as well as a scenario where one line item may require the product level promotion to be applied, while another may not?

Another note is that you may need to call applyDiscounts multiple times within your routine due to the product vs order level promotion calculations.

**Comment by Ivan-Melnyk** [comment]
Hi @GeneralPatten thank for the reply.
The promotions are configured as normally, one for product level discount, another one for order level discount, both percentage (buy 1 product and get 10% off, for both promotions). I've tried to call PromotionMgr.applyDiscounts(discountPlan) several time during the calculate.js, after each product line item and its discount removed, then after all of the product price calculation, and even before tax calculation and the result is the same, not what I'm expecting. I'm expecting to have the Order level discounts applied based on (the discounted product total price minus removed product level discounts), but the result is that the order level discount still calculated based on discounted product level price including all discounts.

**Comment by tparikh2** [comment]
Hi Ivan,

Can you please confirm where did you make this changes, its in calculate.js
or somewhere else?

Also if not in calculate.js then can you try to put this in calculate.js
before apply discount code?

//PromotionMgr.applyDiscounts(basket);
PromotionMgr.applyDiscounts(discountPlan);


Thanks & Regards,
Tushar Parikh.

+91-909-901-3316

**Comment by Ivan-Melnyk** [comment]
Hi @tparikh2 , yes, it's done in calculate.js

**Comment by GeneralPatten** [comment]
@Ivan-Melnyk I looked into this last night and I had the same behavior/result that you are reporting. I would definitely recommend submitting a ticket to SFCC support. Include the details of your promotion setup, as well as a snippet of your code, and the results you are seeing.

Out of curiosity, what are the business rules that require the discount to be removed programmatically? You should be able to meet nearly any condition for applying discounts to a line item using the promotion engine's configuration.

**Comment by Ivan-Melnyk** [comment]
Hi @GeneralPatten thank for your reply, that's what I did already, I've send an email to SFCC support, was not able to request a ticket from partner.salesforce.com since our partnership model has changes a bit, not sure actually why.
If you want to see the business logic and requirement why we need such thing like remove discounts in code - you can look at our repository https://github.com/SalesforceCommerceCloud/link_globale, since it's quite long story.
